File: 66. Plus One/digit.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int n = digits.size();
        for (int i = n - 1; i >= 0; i--) {
            if (digits[i] < 9) {
                digits[i]++; // 0-8 + 1 = 1-9
                return digits; // 0-8 + 1 = 1-9
            }
            digits[i] = 0; // 9 + 1 = 0
        }
        digits[0] = 1; // 9 + 1 = 0
        digits.push_back(0); // 9 + 1 = 0
        return digits;
    }
};

// The solution is based on the following observation:
// 1. If the last digit is not 9, then we can simply add 1 to the last digit and return the result.
// 2. If the last digit is 9, then we need to set it to 0 and propagate the carry to the left.
// 3. If the first digit is 0, then we need to add a new digit 1 to the beginning of the array.

File: 66. Plus One/digit.py
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        num = 0
        for i in range(len(digits)):
            num += digits[i] * 10**(len(digits)-i-1)
        num += 1

        def get_digit(number, i):
            return number // 10**i % 10
        res = []

        for i in range(len(str(num)) - 1, -1, -1):
            res.append(get_digit(num, i))

        return res